compiler/src/dotty/tools/dotc/core/TypeOps.scala

```diff
@@ -659,42 +659,38 @@ object TypeOps:
    */
   private def instantiateToSubType(tp1: NamedType, tp2: Type)(using Context): Type = {
     /** expose abstract type references to their bounds or tvars according to variance */
-    class AbstractTypeMap(maximize: Boolean)(using Context) extends TypeMap {
-      def expose(lo: Type, hi: Type): Type =
-        if (variance == 0)
-          newTypeVar(TypeBounds(lo, hi))
-        else if (variance == 1)
-          if (maximize) hi else lo
-        else
-          if (maximize) lo else hi
+    class ApproximateTypeParams(using Context) extends TypeMap {
+      val boundTypeParams = util.HashMap[TypeRef, TypeVar]()
 
-      def apply(tp: Type): Type = tp match {
+      def apply(tp: Type): Type = tp.dealias match {
         case _: MatchType =>
           tp // break cycles
 
-        case tp: TypeRef if isBounds(tp.underlying) =>
-          val lo = this(tp.info.loBound)
-          val hi = this(tp.info.hiBound)
-          // See tests/patmat/gadt.scala  tests/patmat/exhausting.scala  tests/patmat/t9657.scala
-          val exposed = expose(lo, hi)
-          typr.println(s"$tp exposed to =====> $exposed")
-          exposed
-
-        case AppliedType(tycon: TypeRef, args) if isBounds(tycon.underlying) =>
-          val args2 = args.map(this)
-          val lo = this(tycon.info.loBound).applyIfParameterized(args2)
-          val hi = this(tycon.info.hiBound).applyIfParameterized(args2)
-          val exposed = expose(lo, hi)
-          typr.println(s"$tp exposed to =====> $exposed")
-          exposed
+        case tp: TypeRef if !tp.symbol.isClass =>
+          def lo = apply(tp.underlying.loBound)
+          def hi = apply(tp.underlying.hiBound)
+          boundTypeParams.getOrElseUpdate(tp, newTypeVar(TypeBounds(lo, hi)))
+
+        case AppliedType(tycon: TypeRef, _) if !tycon.dealias.typeSymbol.isClass =>
+          // Type inference cannot handle X[Y] <:< Int
```

Comment:
I'm confused by what is happening here:
- Could you describe in more details what the issue with type inference is here?
- Replacing a type lambda by its result type means that the type now contains references to binders which are not in scope, isn't that problematic?

Reply:
The issue here can be demonstrated by the example:

```Scala
    trait ModuleSig {
      type F[_]
      type U

      trait FooSig {
        type Type = F[U]
        def subst[F[_]](fa: F[Int]): F[Type]
      }

      val Foo: FooSig
    }
    val Module: ModuleSig = new ModuleSig {
      type F[A] = Int

      val Foo: FooSig = new FooSig {
        // type Type = Int
        def subst[F[_]](fa: F[Int]): F[Type] = fa
      }
    }
    type Foo = Module.Foo.Type

    sealed abstract class K[F]
    case object K1 extends K[Int]
    case object K2 extends K[Foo]

    val kv: K[Foo] = Module.Foo.subst[K](K1)
    def test(k: K[Foo]): Unit = k match {
      case K2 => ()
    }
```

We need to tell whether it's possible that `K1 <: K[Foo]`. If yes, we issue a warning; otherwise, no warnings.

- `K1 <: K[Foo]` is possible   <==> 
- `K[Int] <: K[Foo]` is possible   <==>
- `Int <: Foo` is possible   <==>
- `Int <: Module.Foo.Type` is possible

Standard subtyping does not answer whether it is _possible_` S <: T` given some instantiation of the type parameters. Therefore, we replace type parameters in `S` and `T` with tvars, and see if the subtyping is true.

In the example above (if we remove the special case), we will encounter the case `Int <: X[Y]`, where `X` and `Y` are tvars. The subtype checking will simply return false. But depending on the bounds of `X` and `Y`, the subtyping can be true.

As a workaround, we approximate higher-kinded type parameters with the value types that can be instantiated from its bounds.

> Replacing a type lambda by its result type means that the type now contains references to binders which are not in scope, isn't that problematic?

This is indeed a concern. I guess `instantiate` the `HKLambda` with its bounds is a more appropriate way to achieve that. The bounds could be F-Bounds again, we can make it recur for the possibility of F-bounds.

Reply:
Actually referencing the bound TypeParamRef in the HKTypeLambda here is fine, as TypeComparer will recurse on its bounds.